### PR TITLE
Producer Retries, New Message Option Funcs (and more)

### PR DESCRIPTION
**Pull request by deltics**
## Summary by CodeRabbit

- **New Features**
  - Added a mock producer for easier testing.
  - Introduced logging features for custom log handling.
  - Added a new `If` function for conditional message handling.

- **Bug Fixes**
  - Enhanced error handling and state transitions in consumer operations.
  - Improved message creation and manipulation functions.

- **Documentation**
  - Updated README to reflect simplified configuration, reduced boilerplate, and added retries.

- **Tests**
  - Added extensive test coverage for new functionalities and edge cases.
  
- **Refactor**
  - Reorganized test structures for better clarity and coverage.
  - Improved error handling and retry logic in producer methods.

- **Style**
  - Added JSON tags for serialization in logging-related structs.

### CODE REVIEW

File: message.go [hunk 1]
```diff
@@ -2,26 +2,75 @@ package kafka
...
+//		message.Header("key", "value"),
+//		message.JSON(model),
+//	)
+func NewMessage[T comparable](topic T, opts ...MessageOption) Message {
+	opts = append([]MessageOption{message.Topic(topic)}, opts...)
+	msg, err := CreateMessage(opts...)
+	if err != nil {
+		panic(err)
+	}
+	return msg
```
[the remark this answers is not in the document]
Comment (deltics) [reply message.go:69]: NewMessage deliberately panics since the errors that could occur would be programming errors.  If the caller does wish or need to handle any errors they would use the CreateMessage function instead.

File: consumer.go [hunk 2]
```diff
@@ -259,28 +278,109 @@ func (c *consumer) cleanup(ctx context.Context, err error, recovered any) {
...
+		defer c.Unlock()
+
+		if c.waiting++; c.waiting > 1 {
+			return fmt.Errorf("%w: there must be at most ONE Wait() call for a consumer", ErrInvalidOperation)
+		}
+		if c.state < csSubscribed {
+			return ErrConsumerNotStarted
+		}
+		return <-c.stopped
+	}()
```
[the remark this answers is not in the document]
Comment (deltics) [reply consumer.go:308]: In the current implementation, multiple calls to Wait() are always a programming error.

File: consumer.go [hunk 3]
```diff
@@ -181,14 +186,17 @@ func (c *consumer) run(ctx context.Context) {
...
 	}
-	exitLog := logs.Info // upgraded to logs.Error if an error occurs
+	func() {
+		c.Lock()
+		defer c.Unlock()
 
-	logs.Info(ctx, "waiting for messages", exitInfo)
+		c.state = csRunning
+		logs.Info(ctx, "waiting for messages", exitInfo)
+	}()
```
[the remark this answers is not in the document]
Comment (deltics) [reply consumer.go:199]: Provided setState() for simple state changes, such as required by the run() method; run() refactored to use that method.
